src/naive/date.rs

```diff
@@ -597,31 +597,99 @@ impl NaiveDate {
+    /// );
+    /// ```
+    pub fn checked_add_months(self, months: Months) -> Option<Self> {
+        if months.0 == 0 {
+            return Some(self);
+        }
+
+        match months.0 <= core::i32::MAX as u32 {
+            true => self.diff_months(months.0 as i32),
+            false => None,
+        }
+    }
+
+    /// Subtract a duration in [`Months`] from the date
     ///
-    /// A new NaiveDate on the first day of the resulting year & month
-    fn add_months_get_first_day(&self, delta: i32) -> NaiveDate {
-        let zeroed_months = self.month() as i32 - 1; // zero-based for modulo operations
-        let res_months = zeroed_months + delta;
-        let delta_years = if res_months < 0 {
-            if (-res_months) % 12 > 0 {
-                res_months / 12 - 1
-            } else {
-                res_months / 12
+    /// If the day would be out of range for the resulting month, use the last day for that month.
+    ///
+    /// Returns `None` if the resulting date would be out of range.
+    ///
+    /// ```
+    /// # use chrono::{NaiveDate, Months};
+    /// assert_eq!(
+    ///     NaiveDate::from_ymd(2022, 2, 20).checked_sub_months(Months::new(6)),
+    ///     Some(NaiveDate::from_ymd(2021, 8, 20))
+    /// );
+    /// ```
+    pub fn checked_sub_months(self, months: Months) -> Option<Self> {
+        if months.0 == 0 {
+            return Some(self);
+        }
+
+        // Copy `i32::MIN` here so we don't have to do a complicated cast
+        match months.0 <= 2_147_483_648 {
+            true => self.diff_months(-(months.0 as i32)),
+            false => None,
+        }
+    }
+
+    fn diff_months(self, months: i32) -> Option<Self> {
+        let (years, left) = ((months / 12), (months % 12));
+
+        // Determine new year (without taking months into account for now
+
+        let year = if (years > 0 && years > (MAX_YEAR - self.year()))
+            || (years < 0 && years < (MIN_YEAR - self.year()))
+        {
+            return None;
+        } else {
+            self.year() + years
+        };
+
+        // Determine new month
+
+        let month = self.month() as i32 + left;
+        let (year, month) = if month <= 0 {
+            if year == MIN_YEAR {
+                return None;
+            }
+
+            (year - 1, month + 12)
+        } else if month > 12 {
+            if year == MAX_YEAR {
+                return None;
             }
+
+            (year + 1, month - 12)
         } else {
-            res_months / 12
+            (year, month)
         };
-        let res_years = self.year() + delta_years;
-        let res_months = res_months % 12;
-        let res_months = if res_months < 0 { res_months + 12 } else { res_months };
-        NaiveDate::from_ymd(res_years, res_months as u32 + 1, 1)
+
+        // Clamp original day in case new month is shorter
+
+        let flags = YearFlags::from_year(year);
+        let feb_days = if flags.ndays() == 366 { 29 } else { 28 };
```

Comment:
This is a much better implementation, thank you! Had I known about `ndays()` I'd have done things differently. Either way, very happy to have support for adding Months in chrono!